src/Castle.Windsor/MicroKernel/Resolvers/DefaultDependencyResolver.cs

```diff
@@ -424,23 +418,37 @@ private object ResolveFromParameter(CreationContext context, ComponentModel mode
 			}
 		}
 
-		private IHandler TryGetHandlerFromKernel(DependencyModel dependency, CreationContext context)
+		private bool TryGetHandlerFromKernel(DependencyModel dependency, CreationContext context, out IHandler handler)
 		{
 			// we are doing it in two stages because it is likely to be faster to a lookup
 			// by key than a linear search
-			var handler = kernel.LoadHandlerByType(dependency.DependencyKey, dependency.TargetItemType, context.AdditionalArguments);
-			if (handler == null)
+			try
+			{
+				handler = kernel.LoadHandlerByType(dependency.DependencyKey, dependency.TargetItemType, context.AdditionalArguments);
+			}
+			catch (HandlerException)
 			{
-				throw new HandlerException(string.Format("Handler for {0} was not found.", dependency.TargetItemType), null);
+				handler = null;
 			}
+			if (handler == null) return false;
+
 			if (handler.IsBeingResolvedInContext(context) == false)
 			{
-				return handler;
+				return true;
 			}
 
 			// make a best effort to find another one that fit
 
-			var handlers = kernel.GetHandlers(dependency.TargetItemType);
+			IHandler[] handlers;
+			try
+			{
+				handlers = kernel.GetHandlers(dependency.TargetItemType);
+			}
+			catch (HandlerException)
```

Comment:
To make this a valid refactoring, we have to be sure that this method can't throw `HandlerException` indirectly.
If you know that `kernel.GetHandlers` would not throw `HandlerException`, we can remove the `try`. Same for the `try` above.

Reply:
`DefaultNamingSubSystem` doesn't appear to, however a custom `INamingSubSystem` or `IHandlersFilter` may. Sounds like a sensible way to write it.